File: alg_ds_1/queue/queue-2.go
package main

import (
	"errors"
	"io"
	"os"
)

// Задание 5 - функция с обращением очереди - с созданием временного слайса, в который выгружается вся очередь,
// а потом загружается обратно с другого конца - написана для очереди из основного варианта
func (q *Queue[T]) Reverse() {
	if q.size <= 1 {
		return
	}

	stack := make([]T, 0, q.size)
	for q.size > 0 {
		val, _ := q.Dequeue()
		stack = append(stack, val)
	}

	for i := len(stack) - 1; i >= 0; i-- {
		q.Enqueue(stack[i])
	}
}

// Задание 3 - вращение очереди по кругу на N элементов (в любую сторону)
func RotateQueue[T any](q *Queue[T], n int) {
	if q.Size() == 0 {
		return
	}

	n = n % q.Size()
	if n < 0 {
		n += q.Size()
	}

	for i := 0; i < n; i++ {
		if val, err := q.Dequeue(); err == nil {
			q.Enqueue(val)
		}
	}
}

// Stack реализация - из-за особенностей реализации мы получаем сложность O(n)
// для операции извлечения когда выходной стек пуст, потому что придется все элементы перекладывать.
// Такая операция не будет проходить слишком часто, поэтому амортизировання сложность будет ниже, но
// "чистая реализация" выглядит лучше.
// Затрудняюсь представить, зачем может быть нужна такая реализация, кроме демонстрации
type Stack[T any] struct {
	items []T
}

func (s *Stack[T]) Push(item T) {
	s.items = append(s.items, item)
}

func (s *Stack[T]) Pop() (T, error) {
	if len(s.items) == 0 {
		var zero T
		return zero, errors.New("stack is empty")
	}
	item := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return item, nil
}

func (s *Stack[T]) Size() int {
	return len(s.items)
}

// Задание 4 - реализация очереди на двух стеках
type Queue[T any] struct {
	inStack  Stack[T]
	outStack Stack[T]
}

func (q *Queue[T]) Enqueue(item T) {
	q.inStack.Push(item)
}

func (q *Queue[T]) Dequeue() (T, error) {
	// Если выходной стек пуст, перекладываем элементы из входного
	if q.outStack.Size() == 0 {
		for q.inStack.Size() > 0 {
			item, _ := q.inStack.Pop()
			q.outStack.Push(item)
		}
	}

	return q.outStack.Pop()
}

func (q *Queue[T]) Size() int {
	return q.inStack.Size() + q.outStack.Size()
}

// Задание 6 - циклическая очередь через статический массив (то есть нужна проерка полная/неполная)
// через 2 указателя - один указывает на начало очереди, второй - на конец
// если размер очереди становится равным 0, указатели сбрасываются
// Также здесь, как и в обычной круговой очереди, используется остаток деления емкости очереди на индекс для
// зацикливания

type CircularQueue[T any] struct {
	data     []T
	front    int
	rear     int
	size     int
	capacity int
}

func NewCircularQueue[T any](capacity int) *CircularQueue[T] {
	return &CircularQueue[T]{
		data:     make([]T, capacity),
		front:    -1,
		rear:     -1,
		size:     0,
		capacity: capacity,
	}
}

func (q *CircularQueue[T]) IsEmpty() bool {
	return q.size == 0
}

func (q *CircularQueue[T]) IsFull() bool {
	return q.size == q.capacity
}

func (q *CircularQueue[T]) Enqueue(item T) error {
	if q.IsFull() {
		return errors.New("queue is full")
	}

	if q.IsEmpty() {
		q.front = 0
	}

	q.rear = (q.rear + 1) % q.capacity
	q.data[q.rear] = item
	q.size++
	return nil
}

func (q *CircularQueue[T]) Dequeue() (T, error) {
	var zero T
	if q.IsEmpty() {
		return zero, errors.New("queue is empty")
	}

	item := q.data[q.front]
	if q.front == q.rear {
		q.front = -1
		q.rear = -1
	} else {
		q.front = (q.front + 1) % q.capacity
	}
	q.size--
	return item, nil
}

func (q *CircularQueue[T]) Peek() (T, error) {
	var zero T
	if q.IsEmpty() {
		return zero, errors.New("queue is empty")
	}
	return q.data[q.front], nil
}

File: alg_ds_1/python_version/05_queue/queue_2.py
from queue import Queue as BaseQueue


# Задание на курсе: 5
# Задача: 3
# Название: циклическое вращение очереди на m элементов
# Временная сложность: O(n^2), где n — размер очереди
# Пространственная сложность: O(1)
def rotate_queue(queue, m):
    if queue.size() == 0:
        return

    rotations = m % queue.size()

    for _ in range(rotations):
        queue.enqueue(queue.dequeue())


# Задание на курсе: 5
# Задача: 4
# Название: очередь на двух стеках
class TwoStackQueue:

    def __init__(self):
        self.input_stack = []
        self.output_stack = []

    # Временная сложность: амортизированная O(1)
    # Пространственная сложность: O(1)
    def enqueue(self, item):
        self.input_stack.append(item)

    # Временная сложность: O(n), амортизированная O(1)
    # Пространственная сложность: O(1)
    def dequeue(self):
        if self.size() == 0:
            return None

        if len(self.output_stack) > 0:
            return self.output_stack.pop()

        while len(self.input_stack) > 0:
            self.output_stack.append(self.input_stack.pop())

        return self.output_stack.pop()

    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def size(self):
        return len(self.input_stack) + len(self.output_stack)


# Задание на курсе: 5
# Задача: 5
# Название: очередь с обращением элементов
class Queue(BaseQueue):

    # Временная сложность: O(n^2)
    # Пространственная сложность: O(n)
    def reverse(self):
        stack = []

        while self.size() > 0:
            stack.append(self.dequeue())

        while len(stack) > 0:
            self.enqueue(stack.pop())


# Задание на курсе: 5
# Задача: 6
# Название: циклическая очередь в статическом массиве
class CircularQueue:

    def __init__(self, capacity):
        self.capacity = capacity
        self.queue = [None] * capacity
        self.head = 0
        self.tail = 0
        self.count = 0

    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def size(self):
        return self.count

    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def is_full(self):
        return self.count == self.capacity

    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def enqueue(self, item):
        if self.is_full():
            return None

        self.queue[self.tail] = item
        self.tail = (self.tail + 1) % self.capacity
        self.count += 1

    # Временная сложность: O(1)
    # Пространственная сложность: O(1)
    def dequeue(self):
        if self.size() == 0:
            return None

        item = self.queue[self.head]
        self.queue[self.head] = None
        self.head = (self.head + 1) % self.capacity
        self.count -= 1

        return item


#  Рефлексия
# 1. Делал примерно так же, с учетной стоимостью append, но реаллокация по заполненному
# массиву, не из банка.

# 2. Не подумал про разложение на линейном массиве, совсем. Хотя он был на прошлой итерации.
# Получается, что мы должны хранить сам массив и каждое измерение, и далее просто прыгаем по каждому из них,
# Как по блокам.